### PR TITLE
(Apple) Clang: compiler error workaround.

### PR DESCRIPTION
**Pull request by psiha**
Unfortunately - Clang is the new MSVC (tm) :/

(if you can add CI tests for Android and iOS or atleast OSX)

### CODE REVIEW

**Comment by psiha** [comment]
for example - then you'll catch that you also need things like:

`#include <bit>
#if defined( __ANDROID__ ) || defined( __APPLE__ )
#include <boost/integer/static_log2.hpp>
namespace std
{
    template < typename Target >
    constexpr auto bit_cast ( auto const &       source ) noexcept { return __builtin_bit_cast( Target, source );   }
    consteval auto bit_width( auto         const value  ) noexcept { return boost::static_log2< value >::value + 1; }
} // namespace std
#endif`

;)

**Comment by psiha** [comment]
you're also missing <functional> include for unwrap_ref_decay in 'some' place

File: include/kwk/utility/container/shape.hpp [hunk 1]
```diff
@@ -376,9 +406,21 @@ namespace kwk
...
     //!           sub-volume
     //==============================================================================================
+    // has to be defined inline due to (Apple) Clang deficiencies
+    // https://github.com/llvm/llvm-project/issues/58952
+    // but also after of_size
+    //==============================================================================================
     template<typename... Slicers>
-    inline constexpr auto operator()(Slicers const&... s) const noexcept
-    requires( sizeof...(Slicers) <= static_order);
+    constexpr auto operator()(Slicers const&... s ) const noexcept
```
Comment (jfalcou) [inline include/kwk/utility/container/shape.hpp:414]: where is the inline then ?

**Comment by jfalcou** [comment]
The bits part is probably just a libcpp issue. We had it with EVE and we added a clang+libcpp CI instance. We can start with that.

**Comment by jfalcou** [comment]
For the OS X thing, I'll need to see what GA or CircleCI provides

File: include/kwk/utility/container/shape.hpp (hunk 1 above)
Comment (psiha) [reply include/kwk/utility/container/shape.hpp:414]: inline as in within the class (check the llvm issue/godbolt link) ... the inline is superfluous here (for templates and in-class-definition defines)

**Comment by psiha** [comment]
> The bits part is probably just a libcpp issue. We had it with EVE and we added a clang+libcpp CI instance. We can start with that.

Yes it is a libcpp issue - Apple is especially notorious here - as they use 'random internal' (and old) commits for their official compiler and cpp runtime releases - so they (and Android and sometimes even Emscripten) require special testing ('latest Clang' is simply just not enough)

File: include/kwk/utility/container/shape.hpp (hunk 1 above)
Comment (jfalcou) [reply include/kwk/utility/container/shape.hpp:414]: Oh OK I misunderstood. Fine then

**Comment by jfalcou** [comment]
I added #52 to track that

**Review by jfalcou — APPROVED** [review]

**Pull request merged**